src/test/java/com/coveo/pushapiclient/DocumentUploadQueueTest.java

```diff
@@ -0,0 +1,192 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class DocumentUploadQueueTest {
+
+    @Mock
+    private UploadStrategy uploadStrategy;
+
+    @InjectMocks
+    private DocumentUploadQueue queue;
+
+    private AutoCloseable closeable;
+    private DocumentBuilder documentToAdd;
+    private DeleteDocument documentToDelete;
+
+    private int oneMegaByte = 1 * 1024 * 1024;
+
+    private String generateStringFromBytes(int numBytes) {
+        // Check if the number of bytes is valid
+        if (numBytes <= 0) {
+            return "";
+        }
+
+        // Create a byte array with the specified length
+        byte[] bytes = new byte[numBytes];
+
+        // Fill the byte array with a pattern of ASCII characters
+        byte pattern = 65; // ASCII value for 'A'
+        for (int i = 0; i < numBytes; i++) {
+            bytes[i] = pattern;
+        }
+
+        return new String(bytes);
+    }
+
+    private DocumentBuilder generateDocumentFromSize(int numBytes) {
+        return new DocumentBuilder("https://my.document.uri?ref=1",
+                "My bulky document")
+                .withData(generateStringFromBytes(numBytes));
+    }
+
+    @Before
+    public void setup() {
+        String twoMegaByteData = generateStringFromBytes(2 * oneMegaByte);
+
+        documentToAdd = new DocumentBuilder(
+                "https://my.document.uri?ref=1",
+                "My new document")
+                .withData(twoMegaByteData);
+
+        documentToDelete = new DeleteDocument("https://my.document.uri?ref=3");
+
+        closeable = MockitoAnnotations.openMocks(this);
+    }
+
+    @After
+    public void closeService() throws Exception {
+        closeable.close();
+    }
+
+    @Test
+    public void testIsEmpty() throws IOException, InterruptedException {
+        assertTrue(queue.isEmpty());
+    }
+
+    @Test
+    public void testIsNotEmpty() throws IOException, InterruptedException {
+        queue.add(documentToAdd);
+        assertFalse(queue.isEmpty());
+    }
+
+    @Test
+    public void testShouldReturnBatch() throws IOException, InterruptedException {
+        BatchUpdate batchUpdate = new BatchUpdate(
+                new ArrayList<>() {
+                    {
+                        add(documentToAdd);
+                    }
+                }, new ArrayList<>() {
+                    {
+                        add(documentToDelete);
+                    }
+                });
+        queue.add(documentToAdd);
+        queue.add(documentToDelete);
+
+        assertEquals(batchUpdate, queue.getBatch());
+    }
+
+    @Test
+    public void testFlushShouldNotUploadDocumentaWhenRequiredSizeIsNotMet() throws IOException, InterruptedException {
+        queue.add(documentToAdd);
+        queue.add(documentToDelete);
```

Comment:
how do these docs not meet the req size?

Reply:
I will add some more explanation to the test. Essentially, each doc is 2MB large and the batch size limit is 5MB.
So there are still space left before automatically flushing the batch